### PR TITLE
Node to shard control

### PR DESCRIPTION
**Pull request by undertome**
## High Level Overview of Change

This change extends the rpc interface by allowing starting, stopping and polling a nodestore to shardstore database import from a single command.

### Context of Change

This is a new feature that allows users to control database imports with the following rpc command syntax:

` node_to_shard [status|start|stop]`

This change also includes a bug fix from @ximinez that corrects a false positive in the shard unit tests.

### Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [x] Tests (You added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation Updates
- [ ] Release

### CODE REVIEW

**Review by ximinez — CHANGES_REQUESTED** [review]
Fails to build under Windows. Here's the fix: https://github.com/ximinez/rippled/commit/8903fa997fe399013c5b24fce7e0f3196a2ec28e

Consider pulling that lambda into a separate function, since it's duplicated.

**Comment by undertome** [comment]
> Fails to build under Windows. Here's the fix: [ximinez@d4e26b0](https://github.com/ximinez/rippled/commit/d4e26b0f5fcafbb158e49ccc19657ba66cf46c34)
> 
> Consider pulling that lambda into a separate function, since it's duplicated.

Done

**Review by ximinez — CHANGES_REQUESTED** [review]

File: src/ripple/nodestore/impl/DatabaseShardImp.cpp [hunk 1]
```diff
@@ -735,7 +737,27 @@ DatabaseShardImp::stop()
...
+
+            std::this_thread::sleep_for(std::chrono::milliseconds(100));
+            lock.lock();
+        }
+
+        // Calling join while holding the mutex_ without
+        // first making sure that doImportDatabase has
+        // exited could lead to deadlock via the mutex
+        // acquisition that occurs in that function
         databaseImporter_.join();
```
Comment (ximinez) [inline src/ripple/nodestore/impl/DatabaseShardImp.cpp:759]: Because the lock was released and reacquired, we can't be sure that `databaseImporter_` is still `joinable()` here. This is particularly true if you make the change I suggested elsewhere to do `databaseImporter_.detach()` at the end of the thread.
```
if (databaseImporter_.joinable())
    databaseImporter_.join();
```

**Review by ximinez — CHANGES_REQUESTED** [review]

File: src/ripple/nodestore/impl/DatabaseShardImp.cpp [hunk 2]
```diff
@@ -1043,16 +1069,17 @@ DatabaseShardImp::doImportDatabase()
...
 
     {
         std::lock_guard lock(mutex_);
-        if (isStopping())
-            return;
 
-        databaseImportStatus_.reset();
+        if (shouldHalt())
+            return;
     }
```
[the remark this answers is not in the document]
Comment (undertome) [reply src/ripple/nodestore/impl/DatabaseShardImp.cpp:1083]: Not needed. Done

File: src/ripple/rpc/impl/Handler.cpp [hunk 3]
```diff
@@ -107,10 +107,7 @@ Handler const handlerArray[]{
     {"log_level", byRef(&doLogLevel), Role::ADMIN, NO_CONDITION},
     {"logrotate", byRef(&doLogRotate), Role::ADMIN, NO_CONDITION},
     {"manifest", byRef(&doManifest), Role::USER, NO_CONDITION},
-    {"nodetoshard_status",
-     byRef(&doNodeToShardStatus),
-     Role::USER,
-     NO_CONDITION},
+    {"node_to_shard", byRef(&doNodeToShard), Role::USER, NO_CONDITION},
```
[the remark this answers is not in the document]
Comment (undertome) [reply src/ripple/rpc/impl/Handler.cpp:110]: Done

File: src/ripple/rpc/handlers/NodeToShard.cpp [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        ret = shardStore->stopNodeToShard();
+    }
+    else
+    {
+        // Shouldn't happen
+        assert(false);
+        return rpcError(rpcINTERNAL);
+    }
+
+    return ret;
```
[the remark this answers is not in the document]
Comment (undertome) [reply src/ripple/rpc/handlers/NodeToShard.cpp:87]: Well a return is needed at the end because of the return type, so I thought it would be easier to just have all the clauses return from the same point in the function.

File: src/test/rpc/NodeToShardRPC_test.cpp [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+        section.set("path", tempDir.path());
+        section.set("max_historical_shards", "20");
+        section.set("ledgers_per_shard", "256");
+        section.set("earliest_seq", "257");
+        auto& sectionNode = c->section(ConfigSection::nodeDatabase());
+        sectionNode.set("earliest_seq", "257");
+        sectionNode.set("ledgers_per_shard", "256");
+        c->setupControl(true, true, true);
+
+        jtx::Env env(*this, std::move(c));
```
[the remark this answers is not in the document]
Comment (undertome) [reply src/test/rpc/NodeToShardRPC_test.cpp:73]: Done

File: src/test/rpc/NodeToShardRPC_test.cpp [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+        // shard store database import.
+        for (int i = 0; i < env.app().getShardStore()->ledgersPerShard() *
+                 (numberOfShards + 1);
+             ++i)
+        {
+            env.close();
+        }
+
+        auto shardStore = env.app().getShardStore();
+        BEAST_EXPECT(shardStore);
```
[the remark this answers is not in the document]
Comment (undertome) [reply src/test/rpc/NodeToShardRPC_test.cpp:87]: Done

File: src/ripple/nodestore/impl/DatabaseShardImp.cpp [hunk 7]
```diff
@@ -1199,10 +1226,10 @@ DatabaseShardImp::sweep()
 Json::Value
 DatabaseShardImp::getDatabaseImportStatus() const
 {
-    Json::Value ret(Json::objectValue);
-
     if (std::lock_guard lock(mutex_); databaseImportStatus_)
     {
+        Json::Value ret(Json::objectValue);
+
```
[the remark this answers is not in the document]
Comment (undertome) [reply src/ripple/nodestore/impl/DatabaseShardImp.cpp:1232]: Done

File: src/ripple/nodestore/impl/DatabaseShardImp.cpp [hunk 8]
```diff
@@ -1215,11 +1242,56 @@ DatabaseShardImp::getDatabaseImportStatus() const
...
+DatabaseShardImp::startNodeToShard()
+{
+    std::lock_guard lock(mutex_);
+
+    if (!init_)
+        return RPC::make_error(rpcINTERNAL, "Shard store not initialized");
+
+    if (databaseImporter_.joinable())
+        return RPC::make_error(
+            rpcINTERNAL, "Database import already in progress");
```
[the remark this answers is not in the document]
Comment (undertome) [reply src/ripple/nodestore/impl/DatabaseShardImp.cpp:1262]: Done

File: src/ripple/nodestore/impl/DatabaseShardImp.cpp (hunk 1 above)
Comment (undertome) [reply src/ripple/nodestore/impl/DatabaseShardImp.cpp:759]: Done

File: src/ripple/rpc/handlers/NodeToShard.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (ximinez) [reply src/ripple/rpc/handlers/NodeToShard.cpp:87]: If all the if/else clauses `return`, the compiler will know that a separate one isn't needed at the end of the function.

File: src/ripple/nodestore/impl/DatabaseShardImp.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (undertome) [reply src/ripple/nodestore/impl/DatabaseShardImp.cpp:1084]: Removed

File: src/ripple/rpc/handlers/NodeToShard.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (undertome) [reply src/ripple/rpc/handlers/NodeToShard.cpp]: Done

File: src/ripple/nodestore/impl/DatabaseShardImp.cpp [hunk 9]
```diff
@@ -1215,11 +1239,59 @@ DatabaseShardImp::getDatabaseImportStatus() const
             currentShard[jss::storedSeqs] = shard->getStoredSeqs();
 
         ret[jss::currentShard] = currentShard;
+
+        if (haltDatabaseImport_)
+            ret[jss::message] = "Database import initiated...";
```
Comment (ximinez) [inline src/ripple/nodestore/impl/DatabaseShardImp.cpp:1244]: This is shown if the import process is stopping, so the message should be something like "Database import stopping..."
